### PR TITLE
Timespinner: migrate to new options api and correct random

### PR DESCRIPTION
**Pull request by Jarno458**
## What is this fixing or adding?
As stated in the title

## How was this tested?
Many generations, and checking spoiler log

### CODE REVIEW

**Comment by Jarno458** [comment]
Currently still draft, as i am looking into a way to not have everyone's yaml setting fails to be detected and silently generate a full default game

File: worlds/timespinner/__init__.py [hunk 1]
```diff
@@ -75,56 +77,82 @@ def create_items(self) -> None:
...
-        slot_data["BackPyramid"] = self.precalculated_weights.flood_pyramid_back
-        slot_data["CastleMoat"] = self.precalculated_weights.flood_moat
-        slot_data["CastleCourtyard"] = self.precalculated_weights.flood_courtyard
-        slot_data["LakeDesolation"] = self.precalculated_weights.flood_lake_desolation
-        slot_data["DryLakeSerene"] = not self.precalculated_weights.flood_lake_serene
-        slot_data["LakeSereneBridge"] = self.precalculated_weights.flood_lake_serene_bridge
-        slot_data["Lab"] = self.precalculated_weights.flood_lab
-
-        return slot_data
+        return {
```
Comment (alwaysintreble) [inline worlds/timespinner/__init__.py:88]:
```suggestion
		options = self.options.as_dict("start_with_jewelry_box", "downloadable_items", <etc>, casing="pascal")
```

File: worlds/timespinner/__init__.py [hunk 2]
```diff
@@ -177,56 +205,57 @@ def create_item(self, name: str) -> Item:
...
-        if (name == 'Tablet' or name == 'Library Keycard V') and not self.is_option_enabled("DownloadableItems"):
-            item.classification = ItemClassification.filler
-        elif name == 'Oculus Ring' and not self.is_option_enabled("EyeSpy"):
-            item.classification = ItemClassification.filler
-        elif (name == 'Kobo' or name == 'Merchant Crow') and not self.is_option_enabled("GyreArchives"):
-            item.classification = ItemClassification.filler
-        elif name in {"Timeworn Warp Beacon", "Modern Warp Beacon", "Mysterious Warp Beacon"} \
-                and not self.is_option_enabled("UnchainedKeys"):
-            item.classification = ItemClassification.filler
+        if hasattr(self, 'options'): # self.options is not always available in this method
```
Comment (alwaysintreble) [inline worlds/timespinner/__init__.py:208]: This comment shouldn't be true anymore

File: worlds/timespinner/__init__.py [hunk 3]
```diff
@@ -177,56 +205,57 @@ def create_item(self, name: str) -> Item:
...
                 excluded_items.add('Mysterious Warp Beacon')
         else:
             excluded_items.add('Timeworn Warp Beacon')
             excluded_items.add('Modern Warp Beacon')
             excluded_items.add('Mysterious Warp Beacon')
 
-        for item in self.multiworld.precollected_items[self.player]:
-            if item.name not in self.item_name_groups['UseItem']:
-                excluded_items.add(item.name)
+        for item_name in self.options.start_inventory.value.keys():
```
Comment (alwaysintreble) [inline worlds/timespinner/__init__.py:250]: I think the old way was more correct here. I'm not aware of what else might put something into precollected, but if you opt in to start_inventory_from_pool this fails, and I don't trust this be safe.

File: worlds/timespinner/__init__.py [hunk 4]
```diff
@@ -35,33 +35,35 @@ class TimespinnerWorld(World):
...
 
         # in generate_early the start_inventory isnt copied over to precollected_items yet, so we can still modify the options directly
-        if self.multiworld.start_inventory[self.player].value.pop('Meyef', 0) > 0:
-            self.multiworld.StartWithMeyef[self.player].value = self.multiworld.StartWithMeyef[self.player].option_true
-        if self.multiworld.start_inventory[self.player].value.pop('Talaria Attachment', 0) > 0:
-            self.multiworld.QuickSeed[self.player].value = self.multiworld.QuickSeed[self.player].option_true
-        if self.multiworld.start_inventory[self.player].value.pop('Jewelry Box', 0) > 0:
-            self.multiworld.StartWithJewelryBox[self.player].value = self.multiworld.StartWithJewelryBox[self.player].option_true
+        if self.options.start_inventory.value.pop('Meyef', 0) > 0:
+            self.options.start_with_meyef.value = self.options.start_with_meyef.option_true
```
Comment (alwaysintreble) [inline worlds/timespinner/__init__.py:57]: Here also. Should be using the class attribute, not the instance attribute.

**Review by alwaysintreble — COMMENTED** [review]
A few minor suggestions, but seems fine overall. Only 1 or 2 things I actually have a gripe with

File: worlds/timespinner/__init__.py (hunk 1 above)
Comment (Jarno458) [reply worlds/timespinner/__init__.py:88]: Id rather not do this, it wont add much to readability

Comment (alwaysintreble) [reply worlds/timespinner/__init__.py:88]: it's a lot more readable if you put the option names on their own lines

File: worlds/timespinner/__init__.py [hunk 5]
```diff
@@ -35,33 +35,35 @@ class TimespinnerWorld(World):
     Timespinner is a beautiful metroidvania inspired by classic 90s action-platformers.
     Travel back in time to change fate itself. Join timekeeper Lunais on her quest for revenge against the empire that killed her family.
     """
-
-    option_definitions = timespinner_options
+    options_dataclass = BackwardsCompatiableTimespinnerOptions
+    options: BackwardsCompatiableTimespinnerOptions
     game = "Timespinner"
     topology_present = True
     data_version = 12
```
[the remark this answers is not in the document]
Comment (Jarno458) [reply worlds/timespinner/__init__.py]: Got resolved by phar's PR

File: worlds/timespinner/__init__.py (hunk 1 above)
Comment (Jarno458) [reply worlds/timespinner/__init__.py:87]: oke i will keep it the way it is now

File: worlds/timespinner/__init__.py (hunk 2 above)
Comment (Jarno458) [reply worlds/timespinner/__init__.py:207]: I hope your right... as its not an easy scenario to test

File: worlds/timespinner/__init__.py (hunk 3 above)
Comment (Jarno458) [reply worlds/timespinner/__init__.py:249]: this is fine

File: worlds/timespinner/Locations.py [hunk 6]
```diff
@@ -203,7 +202,7 @@ def get_location_datas(world: Optional[MultiWorld], player: Optional[int],
     ]
 
     # 1337156 - 1337170 Downloads
-    if not world or is_option_enabled(world, player, "DownloadableItems"):
+    if not options or options.downloadable_items:
```
Comment (ScipioWright) [inline worlds/timespinner/Locations.py:205]: In what case would `not options` happen?

Comment (Jarno458) [reply worlds/timespinner/Locations.py:205]: When `get_location_datas` is called by `__inti__.py` on line 47, 
`    location_name_to_id = {location.name: location.code for location in get_location_datas(None, None, None)}`

**Review by ScipioWright — APPROVED** [review]
Code looks fine, test gens succeeded.

**Review by agilbert1412 — APPROVED** [review]
my only concern is like the precollected vs start inventory thing, that was already mentioned

File: worlds/timespinner/__init__.py (hunk 3 above)
Comment (NewSoupVi) [reply worlds/timespinner/__init__.py:248]: 3rding that I also don't really understand this change

File: worlds/timespinner/LogicExtensions.py [hunk 7]
```diff
@@ -10,17 +10,18 @@ class TimespinnerLogic:
...
 
-    def __init__(self, world: MultiWorld, player: int, precalculated_weights: PreCalculatedWeights):
+    def __init__(self, player: int, options: Optional[TimespinnerOptions], 
+                 precalculated_weights: Optional[PreCalculatedWeights]):
         self.player = player
 
-        self.flag_specific_keycards = is_option_enabled(world, player, "SpecificKeycards")
-        self.flag_eye_spy = is_option_enabled(world, player, "EyeSpy")
-        self.flag_unchained_keys = is_option_enabled(world, player, "UnchainedKeys")
+        self.flag_specific_keycards = options and options.specific_keycards.value
```
Comment (Exempt-Medic) [inline worlds/timespinner/LogicExtensions.py:22]: I don't believe these need the `.value`

File: worlds/timespinner/__init__.py (hunk 1 above)
Comment (Exempt-Medic) [reply worlds/timespinner/__init__.py:87]: I'm seconding this idea FWIW

File: worlds/timespinner/__init__.py (hunk 3 above)
Comment (Exempt-Medic) [reply worlds/timespinner/__init__.py:248]: 4thing this, the change seems strange. Why was it made?

File: worlds/timespinner/Options.py [hunk 8]
```diff
@@ -388,67 +357,256 @@ class TrapChance(Range):
...
+    """Manual +/-/normal odds for an orb. Put 0 if you don't want a certain nerf or buff to be a possibility. Orbs that
+    you don't specify will roll with 1/1/1 as odds"""
+    visibility = Visibility.none
+
+class HiddenRisingTidesOverrides(RisingTidesOverrides):
+    """Odds for specific areas to be flooded or drained, only has effect when RisingTides is on.
+    Areas that are not specified will roll with the default 33% chance of getting flooded or drained"""
+    visibility = Visibility.none
+
+class HiddenTraps(Traps):
```
Comment (Exempt-Medic) [inline worlds/timespinner/Options.py:417]: I'm wondering why these are fully hidden, these certainly sound like options users would want to actually use

File: worlds/timespinner/Options.py [hunk 9]
```diff
@@ -388,67 +357,256 @@ class TrapChance(Range):
...
+            self.traps == Traps.default:
+            self.traps.value = self.Traps.value
+            has_replaced_options = True
+        if self.DeathLink != DeathLink.default and \
+            self.death_link == DeathLink.default:
+            self.death_link.value = self.DeathLink.value
+            has_replaced_options = True
+
+        if has_replaced_options:
+            logging.warning("Timespinner options where renamed from PasCalCase to snake_case, plz update your yml")
```
Comment (Exempt-Medic) [inline worlds/timespinner/Options.py:612]: End of File newline, also it's far more commonly "yaml" than "yml"
```suggestion
            logging.warning("Timespinner options where renamed from PasCalCase to snake_case, plz update your yaml")
            
```

File: worlds/timespinner/__init__.py (hunk 3 above)
Comment (Jarno458) [reply worlds/timespinner/__init__.py:248]: oke ill revert this one

**Comment by Jarno458** [comment]
I think this is ready for merge now, seems to work fine on my tests

**Review by NewSoupVi — APPROVED** [review]
Looks good now

**Pull request merged**